README
liteAccordion
liteAccordion is a horizontal accordion plugin for jQuery.

Please post any support or feature requests on http://nicolahibbert.com/liteAccordion-v2/
Please do not post support requests on the bug tracker!

Follow me on Twitter (http://twitter.com/nicolahibbert) for the latest liteAccordion development news.

Options:
These are the default settings for the liteAccordion plugin:

containerWidth : 960,                   // fixed (px)
containerHeight : 320,                  // fixed (px)
headerWidth: 48,                        // fixed (px)

activateOn : 'click',                   // click or mouseover
firstSlide : 1,                         // displays slide (n) on page load
slideSpeed : 800,                       // slide animation speed
onTriggerSlide : function() {},         // callback on slide activate
onSlideAnimComplete : function() {},    // callback on slide anim complete

autoPlay : false,                       // automatically cycle through slides
pauseOnHover : false,                   // pause on hover
cycleSpeed : 6000,                      // time between slide cycles
easing : 'swing',                       // custom easing function
                                        
theme : 'basic',                        // basic, dark, light, or colour
rounded : false,                        // square or rounded corners
enumerateSlides : false,                // put numbers on slides 
linkable : false                        // link slides via hash

CHANGELOG
v2.0
    - new methods:
        - play
        - stop
        - next
        - prev
        - destroy
        - debug    

    - rewrote plugin to expose methods
    - refactored core
    - refactored css, styles no longer cascade into slide content divs
    - h2 header now replaceable with any element
    - mouseover activation
    - easing support
    - linkable slides via hashchange
    - two new themes:
        - light
        - colour
    - demo suite available

v1.1.3 06/04/2011
    - IE bug fixes

v1.1.1, v1.1.2 27/03/2011
    - IE bug fixes

v1.1 21/03/2011
    - Added a pause on hover feature

v1.0 02/03/2011
    - First release

test/test.js
module('Instantiation', {
    setup : function() {
        this.test = $('#test').liteAccordion();
    },
    teardown : function() {
        this.test.liteAccordion('destroy');
    }
});

test('jQuery in page', function() {
    strictEqual(typeof jQuery, 'function');
});

test('liteAccordion.js in page', function() {
    strictEqual(typeof $.fn.liteAccordion, 'function');
});

test('liteAccordion.css in page', function() {
    ok($('link[href*="liteaccordion"]', document.head).length);
});

test('DOM element returned', function() {
    strictEqual(typeof this.test, 'object', 'elem is an object');
    strictEqual(this.test[0].nodeType, 1, 'elem is a DOM object'); // instanceof HTMLElement doesn't work in IE7 (it doesn't support DOM L2)
    strictEqual(this.test[0].id, 'test', 'elem has ID of "test"');
});

module('Methods', {
    setup : function() {
        this.test = $('#test').liteAccordion();
        this.debug = this.test.liteAccordion('debug');
    },
    teardown : function() {
        this.test.liteAccordion('destroy');
    }
});

test('Play', function() {
    this.test.liteAccordion('play');
    strictEqual(typeof this.debug.core.playing, 'number', 'On Play, core.playing has a setInterval ID');   
});

test('Play -> Stop', function() {
    this.test.liteAccordion('play');
    this.test.liteAccordion('stop');
    strictEqual(this.debug.core.playing, 0, 'On Stop, core.playing is zero');       
});

test('Play -> Stop -> Play', function() {
    this.test.liteAccordion('play');
    this.test.liteAccordion('stop');
    this.test.liteAccordion('play');
    strictEqual(typeof this.debug.core.playing, 'number', 'On Play, core.playing has a setInterval ID');      
});

test('AutoPlay -> Stop -> Play', function() {
    // destroy instance created in setup
    this.test.liteAccordion('destroy');
    this.test = null;
    this.debug = null;

    // create new instance with autoplay option
    this.test = $('#test').liteAccordion({ autoPlay : true });
    this.debug = this.test.liteAccordion('debug');
    ok(this.debug.settings.autoPlay, 'AutoPlay enabled');
    strictEqual(typeof this.debug.core.playing, 'number', 'On AutoPlay, core.playing has a setInterval ID'); 
    
    // stop autoPlay
    this.test.liteAccordion('stop');
    strictEqual(this.debug.core.playing, 0, 'On Stop, core.playing is zero');

    // manually play
    this.test.liteAccordion('play');
    strictEqual(typeof this.debug.core.playing, 'number', 'On Play, core.playing has a setInterval ID');
});

test('Stop', function() {
    this.test.liteAccordion('stop');
    strictEqual(this.debug.core.playing, 0, 'On Stop, core.playing is zero'); 
});

test('Next', function() {
    strictEqual(this.debug.core.currentSlide, this.debug.settings.firstSlide - 1);    
    this.test.liteAccordion('prev');
    this.test.liteAccordion('next');   
});

test('Next -> Next', function() {
    strictEqual(this.debug.core.currentSlide, this.debug.settings.firstSlide - 1);    
    this.test.liteAccordion('prev');
    this.test.liteAccordion('next');   
});

test('Next -> Prev', function() {
    strictEqual(this.debug.core.currentSlide, this.debug.settings.firstSlide - 1);    
    this.test.liteAccordion('prev');
    this.test.liteAccordion('next');   
});

test('Prev', function() {
    strictEqual(this.debug.core.currentSlide, this.debug.settings.firstSlide - 1);    
    this.test.liteAccordion('prev');
    this.test.liteAccordion('next');   
});

test('Prev -> Prev', function() {
    strictEqual(this.debug.core.currentSlide, this.debug.settings.firstSlide - 1);    
    this.test.liteAccordion('prev');
    this.test.liteAccordion('next');   
});

test('Prev -> Next', function() {
    strictEqual(this.debug.core.currentSlide, this.debug.settings.firstSlide - 1);    
    this.test.liteAccordion('prev');
    this.test.liteAccordion('next');   
});

test('All methods except Debug return DOM element', function() {
    strictEqual(this.test.liteAccordion('play')[0].id, 'test', 'Original DOM element returned');
    strictEqual(this.test.liteAccordion('stop')[0].id, 'test', 'Original DOM element returned');
    strictEqual(this.test.liteAccordion('next')[0].id, 'test', 'Original DOM element returned');
    strictEqual(this.test.liteAccordion('prev')[0].id, 'test', 'Original DOM element returned');
    strictEqual(this.test.liteAccordion('destroy')[0].id, 'test', 'Original DOM element returned');
    strictEqual(this.test.liteAccordion()[0].id, 'test', 'Original DOM element returned'); // init needed for teardown                   
    strictEqual(typeof this.debug, 'object', 'Debug returns an object');   
    ok(this.debug.settings, 'Debug has a settings property');
});



    // strictEqual(this.debug.core.currentSlide, this.debug.settings.firstSlide - 1);
/*
methods
    play
        autoplay, stop, prev, next, play ->check index      

                

            
    next
        next * x -> check current index
        next -> prev -> check index

            
    prev
        prev * rand x -> check current index
        prev -> next -> prev -> check index
        
    
    destroy
        destroys all styles
        destroys events on header and window (hashchange)



TEST:
    autoplay -> stop -> hashchange = correct slide
    autoplay -> pauseonhover = next slide is correct slide
    linkable on load
    linkable on hashchange
    autoplay -> stop -> hashchange
    autoplay -> pauseon hover -> hashchange 
    
        


*/
module('Core', {
    setup : function() {
        this.test = $('#test').liteAccordion();
        this.debug = this.test.liteAccordion('debug');
    },
    teardown : function() {
        this.test.liteAccordion('destroy');
    }
});

test('Play counter on init', function() {
    strictEqual(this.debug.core.playing, 0, 'core.playing is zero');    
});

test('CurrentSlide set on init', function() {
    strictEqual(this.debug.core.currentSlide, this.debug.settings.firstSlide - 1); // firstSlide isn't zero indexed
});


/*
core
    setStyles
    bindEvents
        test clicks!
    linkable    
    currentSlide
    nextSlide
    playing
    getSlidePositions *rewrite
    groupSlides *rewrite
    ungroupSlides *rewrite
    triggerSlide
    init
*/






js/demosuite.js
;(function($) {
    
    var demo = $('#demo'),
        selects = $('.options select'),
        easing = $('#easing'),
        easingFn,
        outputToggle = $('.output h2 a'),
        resetBtn = $('#reset'),
        options,
        update = function() {
            var defaults = demo.liteAccordion('debug').defaults;
            
            // reset 'global' options obj
            options = {};

            selects.each(function() { 
                var current = this.value,
                    value;

                if (!isNaN(parseInt(current, 10))) {
                    value = parseInt(current, 10); 
                } else if (current === 'custom') {
                    value = parseInt($(this).next().val(), 10);
                } else if (current === 'true') {
                    value = true;
                } else if (current === 'false') {
                    value = false;
                } else {
                    value = current;
                }
                
                if (defaults[this.name] !== value) {
                    options[this.name] = value; 
                }
                        
            });                 

            demo.liteAccordion('destroy').liteAccordion(options);

            $('.output textarea').text('$("#yourAccordion").liteAccordion(' + JSON.stringify(options) + ');');
            
        },
        reset = function() {
            options = {};
            
            selects.each(function() {
                $(this).children().eq(0).attr('selected', true);
            });

            $('.options').find('input[type=text]').remove();
            easing.find('option[value=swing]').attr('selected', true);
            
            demo.liteAccordion('destroy');
            demo.liteAccordion();
            
            $('.output textarea').text('$("#yourAccordion").liteAccordion();');
        };
        
    // create easing select options from plugin
    for (easingFn in $.easing) {
        if (easingFn !== 'def') {
            easing.append($('<option value="' + easingFn + '">' + easingFn +'</option>')); // should use a frag, cba tho.                    
        }
    }
                
    easing.find('option[value=swing]').attr('selected', true);
    
    // init accordion
    demo.liteAccordion({ 
        onTriggerSlide : function() { console.log('trigger'); }, 
        onSlideAnimComplete : function() { console.log('complete'); },
        linkable : true,
        autoPlay : true, 
        cycleSpeed : 2000,
        pauseOnHover : true
    });

    // get new options on change event
    selects.change(function() {     
        this.value === 'custom' ? $(this).after('<input type="text" />') : update();
    });
    
    // new opts on enter
    $(window).keyup(function(e) {
        if (e.keyCode === 13) update();
    });
    
    // reset
    resetBtn.click(function(e) {
        reset();
        e.preventDefault();
    });
    
    // show/hide code output                        
    outputToggle.click(function(e) {
        $(this).parent().next().slideToggle();
        e.preventDefault();
    });

})(jQuery);

liteAccordion.jQuery.js
/*************************************************
*
*	project:  liteAccordion - horizontal accordion plugin for jQuery
*	author:   Nicola Hibbert
*	url:	  http://nicolahibbert.com/horizontal-accordion-jquery-plugin
*	demo:	  http://www.nicolahibbert.com/demo/liteAccordion
*
*	Copyright (c) 2010-2011 Nicola Hibbert
*	License: MIT
*
/*************************************************/
;(function($) {
	
	$.fn.liteAccordion = function(options) {
		
		// defaults
		var defaults = {
			containerWidth : 960,
			containerHeight : 320,
			headerWidth : 48,
			
			firstSlide : 1, 
			activateOn : 'click', // mouseover
			onActivate : function() {},
			slideSpeed : 800,
			slideCallback : function() {},			
			
			autoPlay : false, 
			cycleSpeed : 5000,
			pauseOnHover : false,

			theme : 'basic', // basic, light, dark
			roundedCorners : false
		},
		
		// merge defaults with options in new settings object				
			settings = $.extend({}, defaults, options),
	
		// define key variables
			$accordion = this,
			$slides = $accordion.find('li'),
			slideLen = $slides.length,
			slideWidth = settings.containerWidth - (slideLen * settings.headerWidth),
			$header = $slides.children('h2');
		
		// set heights, widths, theme & corner style
		$accordion
			.height(settings.containerHeight)
			.width(settings.containerWidth)
			.addClass((settings.theme === 'dark') ? 'dark' : (settings.theme === 'light') ? 'light' : 'basic')
			.addClass(settings.roundedCorners && 'rounded');
		
		$header
			.width(settings.containerHeight)
			.height(settings.headerWidth);
		
		// set initial positions for each slide
		$header.each(function(index) {
			var $this = $(this),
				left = index * settings.headerWidth;
				
				if (index >= settings.firstSlide) {
					left += slideWidth;
				} 
				
				$this
					.css('left', left)
					.next()
						.width(slideWidth + settings.headerWidth)
						.css('left', left);
		});
		
		// bind event handler for activating slides
		$header.bind(settings.activateOn, function() {
			var $this = $(this),
				index = $header.index($this),
				pos = {
					left : index * settings.headerWidth,
					right : index * settings.headerWidth + slideWidth
				}, 
				newPos,
				$group = utils.getGroup.call(this, pos, index); 

			settings.onActivate.call($accordion);

			if (this.offsetLeft === pos.left) {
				newPos = slideWidth;
			} else if (this.offsetLeft === pos.right) {
				newPos = -slideWidth;
			}

			$group
				.add($group.next())
				.animate({
					left : '+=' + newPos
				}, settings.slideSpeed, settings.slideCallback);	
		});
	
		// core utility and animation methods
		var utils = {
			getGroup : function(pos, index) {		
				var $parent = $(this).parent();
				
				if (this.offsetLeft === pos.left) {
					return $parent.nextAll().children(':first-child').filter(function() { return this.offsetLeft === $header.index(this) * settings.headerWidth });
				} else if (this.offsetLeft === pos.right) {
					return $parent.add($parent.prevAll()).children(':first-child').filter(function() { return this.offsetLeft === slideWidth + ($header.index(this) * settings.headerWidth) });	
				}
			},
			next : function() {
				// or prev
				this.parent().next().children(':first-child').trigger('click');
			},
			play : function() {
				if (!settings.autoPlay) return;
				
				// get all slide headers
				// get first active slide
				// trigger next after x amount of time
				
			}
		}	
		
	}; // end plugin

	// setTimeout(function() { $('h2:eq(3)').trigger('click') }, 1000);
	
})(jQuery);

js/liteaccordion.jquery.js
/*************************************************!
*
*   project:    liteAccordion - a horizontal accordion plugin for jQuery
*   author:     Nicola Hibbert
*   url:        http://liteaccordion.nicolahibbert.com
*   demo:       http://liteaccordion.nicolahibbert.com/examples
*
*   Version:    2.0a
*   Copyright:  (c) 2010-2011 Nicola Hibbert
*
**************************************************/

;(function($) {
    
    var LiteAccordion = function(elem, options) {
        
        var defaults = {
            containerWidth : 960,                   // fixed (px)
            containerHeight : 320,                  // fixed (px)
            headerWidth: 48,                        // fixed (px)
                                                    
            activateOn : 'click',                   // click or mouseover
            firstSlide : 1,                         // displays slide (n) on page load
            slideSpeed : 800,                       // slide animation speed
            onTriggerSlide : function() {},         // callback on slide activate
            onSlideAnimComplete : function() {},    // callback on slide anim complete

            autoPlay : false,                       // automatically cycle through slides
            pauseOnHover : false,                   // pause on hover
            cycleSpeed : 6000,                      // time between slide cycles
            easing : 'swing',                       // custom easing function
                                                    
            theme : 'basic',                        // basic, dark, light, or stitch
            rounded : false,                        // square or rounded corners
            enumerateSlides : false,                // put numbers on slides 
            linkable : false                        // link slides via hash, *not supported in IE7
        },

        // merge defaults with options in new settings object   
            settings = $.extend({}, defaults, options),
        
        // 'globals'
            slides = elem.children('ol').children('li'),
            header = slides.children(':first-child'),
            slideLen = slides.length,
            slideWidth = settings.containerWidth - slideLen * settings.headerWidth, 

        // public methods    
            methods = {
                    
                // start elem animation
                play : function(index) {
                    var next = core.nextSlide(index && index);

                    if (core.playing) return;

                    // start autoplay
                    core.playing = setInterval(function() {
                        header.eq(next()).trigger('click.liteAccordion');
                    }, settings.cycleSpeed);
                },
            
                // stop elem animation
                stop : function() {
                    clearInterval(core.playing);
                    core.playing = 0;
                },

                // trigger next slide
                next : function() {
                    methods.stop();
                    header.eq(core.currentSlide + 1).trigger('click.liteAccordion');
                },

                // trigger previous slide
                prev : function() {
                    methods.stop();
                    header.eq(core.currentSlide - 1).trigger('click.liteAccordion');  
                },
                
                // destroy plugin instance
                destroy : function() {                    
                    // stop autoplay
                    methods.stop();

                    // remove hashchange event bound to window
                    $(window).unbind('.liteAccordion');

                    // remove generated styles, classes, data, events
                    elem
                        .attr('style', '')
                        .removeClass('accordion basic dark light stitch')
                        .removeData('liteAccordion')
                        .unbind('.liteAccordion')
                        .find('li > :first-child')
                        .unbind('.liteAccordion')
                        .filter('.selected')
                        .removeClass('selected')
                        .end()
                        .find('b')
                        .remove();
                        
                    slides
                        .removeClass('slide')
                        .children()
                        .attr('style', '');
                },

                // poke around the internals (NOT CHAINABLE)
                debug : function() {
                    return {
                        elem : elem,
                        defaults : defaults,
                        settings : settings,
                        methods : methods,
                        core : core
                    };
                }       
            },

        // core utility and animation methods
            core = {
        
                // set style properties
                setStyles : function() {                    
                    // set container heights, widths, theme & corner style
                    elem
                        .width(settings.containerWidth)
                        .height(settings.containerHeight)
                        .addClass('accordion')
                        .addClass(settings.rounded && 'rounded')                  
                        .addClass(settings.theme);
                        
                    // set tab width, height and selected class
                    slides
                        .addClass('slide')
                        .children(':first-child')
                        .width(settings.containerHeight)
                        .height(settings.headerWidth)
                        .eq(settings.firstSlide - 1)
                        .addClass('selected');

                    // compensate for borders on 'light' theme
                    if (settings.theme === 'light') slideWidth -= parseInt(elem.children('ol').css('borderTopWidth'), 10);

                    // set initial positions for each slide             
                    header.each(function(index) {
                        var $this = $(this),
                            left = index * settings.headerWidth;
                            
                        if (index >= settings.firstSlide) left += slideWidth;

                        $this
                            .css('left', left)
                            .next()
                                .width(slideWidth)
                                .css({ left : left, paddingLeft : settings.headerWidth });

                        // add number to bottom of tab
                        settings.enumerateSlides && $this.append('<b>' + (index + 1) + '</b>');

                    });

                    // ie css fixes
                    if ($.browser.msie) core.ieTransformTest();
                },

                // bind click and mouseover events
                bindEvents : function() {                                        
                    if (settings.activateOn === 'click') {
                        header.bind('click.liteAccordion', core.triggerSlide);
                    } else if (settings.activateOn === 'mouseover') {
                        header.bind({
                            'mouseover.liteAccordion' : core.triggerSlide,
                            'click.liteAccordion' : core.triggerSlide                          
                        });
                    }
                    
                    // pause on hover (can't use custom events with $.hover())      
                    if (settings.pauseOnHover && settings.autoPlay) {
                        elem.bind('mouseover.liteAccordion', function() {
                            core.playing && methods.stop();
                        }).bind('mouseout.liteAccordion', function() {
                            !core.playing && methods.play(core.currentSlide);
                        });
                    } 
                },
                
                linkable : function() {
                    var cacheSlideNames = (function() {
                        var slideNames = [];

                        slides.each(function() {
                            slideNames.push(($(this).attr('name')).toLowerCase());
                        });

                        // memoize
                        return cacheSlideNames = slideNames;                        
                    })();
                    
                    var triggerHash = function(e) {
                        var index;
                        
                        if (e.type === 'load' && !window.location.hash) return;
                        if (e.type === 'hashchange' && core.playing) return;
                        
                        index = $.inArray((window.location.hash.slice(1)).toLowerCase(), cacheSlideNames);
                        if (index > -1 && index < cacheSlideNames.length) header.eq(index).trigger('click.liteAccordion');
                    };

                    $(window).bind({
                        'hashchange.liteAccordion' : triggerHash,
                        'load.liteAccordion' : triggerHash
                    });
                },
                
                // counter for autoPlay (zero index firstSlide on init)
                currentSlide : settings.firstSlide - 1,             

                // next slide index
                nextSlide : function(index) {
                    var next = index + 1 || core.currentSlide + 1;

                    // closure
                    return function() {
                        return next++ % slideLen;
                    };
                },  
    
                // holds interval counter
                playing : 0,
                
                // animates left and right groups of slides
                // side: denotes left side
                animSlideGroup : function(index, next, side) {
                    var filterExpr = side ? ':lt(' + (index + 1) + ')' : ':gt(' + index + ')';

                    slides
                        .filter(filterExpr)
                        .each(function() {
                            var $this = $(this),
                                slideIndex = slides.index($this);
                                
                            $this
                                .children()
                                .stop(true)
                                .animate({
                                    left : (side ? 0 : slideWidth) + slideIndex * settings.headerWidth
                                }, 
                                    settings.slideSpeed, 
                                    settings.easing,
                                    function() { 
                                        // flag ensures that fn is only called one time per triggerSlide
                                        if (!core.slideAnimCompleteFlag) {
                                            settings.onSlideAnimComplete.call(next);
                                            core.slideAnimCompleteFlag = true;
                                        }
                                    });                                     
                        });
                },
                
                slideAnimCompleteFlag : false,
                
                // trigger slide animation
                triggerSlide : function(e) {
                    var $this = $(this),
                        index = header.index($this),
                        next = $this.next();                          
                                                                   
                    // update core.currentSlide
                    core.currentSlide = index;
                    
                    // reset onSlideAnimComplete callback flag
                    core.slideAnimCompleteFlag = false;

                    // remove, then add selected class
                    header.removeClass('selected').filter($this).addClass('selected');               
                 
                    // reset current slide index in core.nextSlide closure
                    if (e.originalEvent && settings.autoPlay) {
                        methods.stop();
                        methods.play(index);
                    }
                    
                    // set location.hash
                    if (settings.linkable) window.location.hash = $this.parent().attr('name');

                    // trigger callback in context of sibling div
                    settings.onTriggerSlide.call(next);

                    // animate left & right groups
                    core.animSlideGroup(index, next, true);
                    core.animSlideGroup(index, next);
                },
                
                ieTransformTest : function() {
                    var div = document.createElement('div');
                    
                    typeof div.style.msTransform === 'string' && elem.addClass('ie9');                
                    div = null;
                },
                
                init : function() {
                    core.setStyles();
                    core.bindEvents();

                    // check slide speed is not faster than cycle speed
                    if (settings.cycleSpeed < settings.slideSpeed) settings.cycleSpeed = settings.slideSpeed;

                    // init hash links
                    if (settings.linkable && 'onhashchange' in window) core.linkable();

                    // init autoplay
                    settings.autoPlay && methods.play();
                }
            };

        // init plugin
        core.init();

        // expose methods
        return methods;
       
    };

    $.fn.liteAccordion = function(method) {
        var elem = this,
            instance = elem.data('liteAccordion');

        // if creating a new instance
        if (typeof method === 'object' || !method) {
            return elem.each(function() {
                var liteAccordion;
    
                // if plugin already instantiated, return
                if (instance) return;

                // otherwise create a new instance
                liteAccordion = new LiteAccordion(elem, method);
                elem.data('liteAccordion', liteAccordion);
            });

        // otherwise, call method on current instance
        } else if (typeof method === 'string' && instance[method]) {
            // debug method isn't chainable b/c we need the debug object to be returned
            if (method === 'debug') {
                return instance[method].call(elem);
            } else { // the rest of the methods are chainable though
                instance[method].call(elem);
                return elem;                
            }
        }
    };

})(jQuery);